🚀 3단계 - 기능 우선 패키지 구성하기

writer0713 opened the pull request:
안녕하세요, 리뷰어님. 
이번 과제는 `Bounded Context` 를 기준으로 패키지 구조를 수정하는 것으로 이해했습니다.

저번 과제에서 `Domain Modeling` 을 하면서 아래 내용들을 정리할 수 있었습니다.
- 어떤 `sub domain` 이 있는지
- 어떤 `bounded context` 가 있는지
- 어떤 `domain event` 가 있는지

이번 과제에서는 저번 `Domain Modeling` 에서 정리된 것을 기준으로 `Bounded Context` 중심의 패키지 구조로 리팩토링(?) 해보았습니다.

아래와 같이 `package` 를 구성했습니다.

```sh
├── Application.java
├── common
│   └── infra
│       └── PurgomalumClient.java
├── menu
│   ├── application
│   │   ├── MenuGroupService.java
│   │   └── MenuService.java
│   ├── domain
│   │   ├── JpaMenuGroupRepository.java
│   │   ├── JpaMenuRepository.java
│   │   ├── Menu.java
│   │   ├── MenuGroup.java
│   │   ├── MenuGroupRepository.java
│   │   ├── MenuProduct.java
│   │   └── MenuRepository.java
│   ├── infra
│   │   └── DefaultPurgomalumClient.java
│   └── ui
│       ├── MenuGroupRestController.java
│       └── MenuRestController.java
├── order
│   ├── application
│   │   ├── OrderService.java
│   │   └── OrderTableService.java
│   ├── domain
│   │   ├── JpaOrderRepository.java
│   │   ├── JpaOrderTableRepository.java
│   │   ├── Order.java
│   │   ├── OrderLineItem.java
│   │   ├── OrderRepository.java
│   │   ├── OrderStatus.java
│   │   ├── OrderTable.java
│   │   ├── OrderTableRepository.java
│   │   └── OrderType.java
│   ├── infra
│   │   ├── DefaultKitchenridersClient.java
│   │   └── KitchenridersClient.java
│   └── ui
│       ├── OrderRestController.java
│       └── OrderTableRestController.java
└── product
    ├── application
    │   └── ProductService.java
    ├── domain
    │   ├── JpaProductRepository.java
    │   ├── Product.java
    │   └── ProductRepository.java
    ├── infra
    │   └── DefaultPurgomalumClient.java
    └── ui
        └── ProductRestController.java
```

writer0713 committed:
refactor: BOUNDED CONTEXT 중심의 패키지 구조로 리팩터링
- 'BOUNDED CONTEXT' 는 'Product', 'Menu', 'Order' 세가지로 분류
- PurgomalumClient 의 경우, 'menu', 'product' 두 context 에서 사용되어 interface 를 common/infra 패키지에 놓고 구현체는 각 context 의 infra 패키지에 위치시킴
- KitchenridersClient 의 경우, 'order' context 에서만 사용되어 order/infra 패키지에 위치시킴

HackSung commented:
정환님이 생각하시는 바운디드 컨텍스트의 정의는 어떻게 될까요? 🤔 
서브도메인과 바운디드 컨텍스트는 같은 개념일까요? 아니면 차이가 있을까요?

지난 미션의 모델링에서 주문 유형별로 구분하여 행위와 속성을 잘 정의해 주셨는데요.
같은 주문이라 하더라도 주문 유형에 따라 진행되는 프로세스나 조건이 다르다는 것을 알게 되셨는데
모델링과 같이 주문 서브도메인을 수업시간에 배운 주문 유형별 바운디드 컨텍스트를 적용하여 패키지를 세분화 해보면 어떨까요?

HackSung commented on src/main/java/kitchenpos/common/infra/PurgomalumClient.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.infra;
+package kitchenpos.common.infra;
```
`PurgomalumClient`인터페이스는 외부 시스템과의 통신을 추상화하여 정의했지만 
어떤 기술을 사용하여 외부 시스템과 통신하는지, 그 통신이 어떻게 이루어지는지에 대한 세부사항은 명시되지 않았는데 
infra계층에 두는게 맞을까요?
정환님이 생각하시는 인프라스트럭쳐 계층에 대한 정의가 궁금합니다. 🤔 

참고로 제가 생각하는 인프라스트럭처 계층의 정의는 다음과 같습니다.
기술적인 세부사항과 관련된 코드를 포함하는 계층으로
예를 들면 데이터베이스 연동, 네트워크 통신, 파일 시스템 접근 등이 있을 것 같습니다.
이 때 도메인 계층이나 애플리케이션 계층이 기술적인 세부 사항을 의존하지 않도록 하는게 중요한데
 `PurgomalumClient` 인터페이스가 그 역할을 담당할 수 있을 것 같습니다.

writer0713 replied:
`PurgomalumClient` 인터페이스는 `application 패키지` 에 두고 구현체인 `DefaultPurgomalumClient` 을 `infra 패키지` 에 두도록 수정하였습니다. :)

writer0713 commented:
@HackSung 
pr 리뷰 요청드리고 몇 시간정도 고민하다가 아래와 같은 궁금증이 생겨 적으려고 하다보니 비슷한 부분에 대해 리뷰를 주셨네요 :) 

## 궁금한점

### 1. `common` 패키지 사용에 대해

가능하면 `common` 패키지를 안 만들고 각 bounded-context 패키지 내부로 위치시키려고 했으나 
(고민 고민하다가 결국) `PurgomalumClient` 의 경우, `공통 interface` 를 `common/infra` 패키지에 놓고 구현체를 `menu` 와 `product` context 의 `infra` 패키지에 위치시켰습니다.
`common` 패키지를 사용하는 것에 대해 어떻게 생각하시는지 궁금합니다.

### 2. 세부적인 `bounded context` 대로 나뉘어지지 않은것 같다는 생각

사실 저번 `modeling` 을 통해 정리된 `bounded context` 는 아래와 같다고 생각합니다.
```
- product
- menu
- menuGroup
- deliveryOrder
- eatInOrder
- takeoutOrder
- orderTable
```
그런데, 기존 `계층 구조 패키지` 에 있는 코드들이 위와 같은 `bounded context` 로 나뉘어져 있지 않다보니 `bounded context` 보다는 `sub domain` 에 맞춰서 패키지가 나뉘어진 느낌이 듭니다.
- common
- `product` (하위 도메인)
- `menu` (하위 도메인)
  - menu_group (x)
- `order` (하위 도메인)
  - delivery_order (x)
  - eat_in_order (x)
  - take_out_order (x)
  - order_table (x)

제가 생각한게 맞는지 궁금합니다.

추가로, 위에서 말씀드린 `bounded context` 들로 패키지를 만들면 기존에 코드들은 새로 만든 패키지에 옮기기가 애매해지는것 같은데요.
우선 패키지만 생성하고 코드들은 그대로 두면 될까요? 아니면, 패키지에 맞게 코드들도 수정을 해야할까요? (ex. order => delivery_order / eat_in_order / take_out_order)

writer0713 committed:
refactor: bounded-context 기준으로 패키지 수정
- 처음 수정했던 패키지는 '하위 도메인' 기준으로 패키지 수정
- 좀 더 세분화하여 'bounded-context' 기준으로 패키지 수정

writer0713 commented:
@HackSung 
말씀하신대로 좀 더 세분화하여 `bounded context` 로 나눠서 다시 정리해봤습니다.
테스트 코드가 잘 동작하도록 하기 위해서 기존 `order` 패키지는 지우지 않고 유지한채로 `delivery_order`, `take_out_order`, `eat_in_order` 세가지 패키지를 추가했습니다.

```java
├── Application.java
├── common
│   └── infra
│       ├── DefaultPurgomalumClient.java
│       └── PurgomalumClient.java
├── delivery_order
│   ├── application
│   ├── domain
│   │   ├── DeliveryOrder.java
│   │   ├── DeliveryOrderRepository.java
│   │   └── JpaDeliveryOrderRepository.java
│   ├── infra
│   └── ui
├── eat_in_order
│   ├── application
│   ├── domain
│   │   ├── EatInOrder.java
│   │   ├── EatInOrderRepository.java
│   │   └── JpaEatInOrderRepository.java
│   ├── infra
│   └── ui
├── menu
│   ├── application
│   │   └── MenuService.java
│   ├── domain
│   │   ├── JpaMenuRepository.java
│   │   ├── Menu.java
│   │   ├── MenuProduct.java
│   │   └── MenuRepository.java
│   ├── infra
│   └── ui
│       └── MenuRestController.java
├── menu_group
│   ├── application
│   │   └── MenuGroupService.java
│   ├── domain
│   │   ├── JpaMenuGroupRepository.java
│   │   ├── MenuGroup.java
│   │   └── MenuGroupRepository.java
│   ├── infra
│   └── ui
│       └── MenuGroupRestController.java
├── order
│   ├── application
│   │   └── OrderService.java
│   ├── domain
│   │   ├── JpaOrderRepository.java
│   │   ├── Order.java
│   │   ├── OrderLineItem.java
│   │   ├── OrderRepository.java
│   │   ├── OrderStatus.java
│   │   └── OrderType.java
│   └── ui
│       └── OrderRestController.java
├── order_table
│   ├── application
│   │   └── OrderTableService.java
│   ├── domain
│   │   ├── JpaOrderTableRepository.java
│   │   ├── OrderTable.java
│   │   └── OrderTableRepository.java
│   ├── infra
│   └── ui
│       └── OrderTableRestController.java
├── product
│   ├── application
│   │   └── ProductService.java
│   ├── domain
│   │   ├── JpaProductRepository.java
│   │   ├── Product.java
│   │   └── ProductRepository.java
│   ├── infra
│   └── ui
│       └── ProductRestController.java
└── take_out_order
    ├── application
    ├── domain
    │   ├── JpaTakeOutOrderRepository.java
    │   ├── TakeOutOrder.java
    │   └── TakeOutOrderRepository.java
    ├── infra
    │   ├── DefaultKitchenridersClient.java
    │   └── KitchenridersClient.java
    └── ui

```

## 궁금한점 

### 1. 공통으로 사용되는 `OrderStatus` / `OrderType` / `OrderLineItem`
`bounded context` 가 나뉘면 각 `bounded context` 에서는 중복이 되더라도 자기만의 객체를 가져야 한다고 이해했었는데요.
맞을까요? 현재는 `order` 패키지에 있는 객체들을 import 해서 사용중입니다.

### 2. `order` 패키지
`bounded context` 대로 나뉜다면 사실상 `order` 패키지는 삭제가 되야 맞을거 같다고 생각이 드는데요.
그리고, `OrderService` 에 있는 로직들이 각 `bounded context` 로 나뉘어져 리팩토링 되야하는것 같습니다.
맞을까요? 현재는 테스트 코드가 성공하는지 확인하기 위해 남겨놨습니다.

HackSung commented on src/main/java/kitchenpos/delivery_order/domain/JpaDeliveryOrderRepository.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.UUID;
+
+public interface JpaDeliveryOrderRepository extends DeliveryOrderRepository, JpaRepository<DeliveryOrder, UUID> {
```
`Jpa...Repository` 이름 패턴으로 작명된 인터페이스들은 모두 `JpaRepository`를 확장하여 JPA라는 특정 기술에 의존하고 있는데 도메인 계층에 위치하는게 맞을까요? 🤔

writer0713 replied:
`...Repository` 는 `domain 패키지` 에 두고 `Jpa...Repository` 를 `infra 패키지` 두도록 수정했습니다 :)

HackSung commented:
> common 패키지 사용에 대해

여러 도메인이 사용하는 객체나 패키지를 공통으로 관리할 지, 
중복되더라도 나중에 수정하기 어려울 수 있으니 처음부터 각각 관리할 지는 장단점이 서로 있기에 늘 고민되는 부분이여서
팀내 정해진 룰이나 개인 선호도에 따라 정해지는 것 같습니다.
DDD에서는 가능한 각 바운디드 컨텍스트가 독립적이고 자기 완결적인게 좋다고 하지만
실제 개발에서는 공통 코드나 도메인을 별도의 공통 패키지로 분리하는 경우가 많은 것 같습니다. 
저는 공통 패키지를 사용하는 것이 DDD의 원칙을 위반한다기 보다 유연성을 제공하는 방법이라고 생각하며
대신에 공통패키지가 특정 컨텍스트의 도메인 모델에 의존하지 않도록 잘 관리하는게 중요할 것 같습니다.

> 우선 패키지만 생성하고 코드들은 그대로 두면 될까요? 아니면, 패키지에 맞게 코드들도 수정을 해야할까요? (ex. order => delivery_order / eat_in_order / take_out_order)

이미 패키지에 맞게 수정하셨군요. 저도 이 방법이 좋아 보입니다. 👍 

> 공통으로 사용되는 OrderStatus / OrderType / OrderLineItem

OrderStatus의 경우는 주문 유형별로 가질 수 있는 상태가 다르기 때문에 해당 주문 유형별로 관련있는 상태로만 구성된 
EatInOrderStatus, DeliveryOrderStatus, TakeoutOrderStatus로 정의하는게 좋아 보입니다.
반면에 OrderType과 OrderLineItem은 모든 주문 유형에서 동일하게 사용하니 공통으로 가져가도 좋을 것 같구요.
참고로 저는 예전에 미션할 때 order를 Parent 패키지로 구성하고 주문 공통과 주문유형별 바운디드 컨테스트를 하위 패키지로 구성하기 했지만 
이것도 한가지 방법일 뿐 정답이 있는게 아니니 정환님 생각대로 구성하시면 될 것 같습니다.

> order 패키지

네 정환님이 생각하시는 것처럼 저도 order 패키지를 삭제하고 테스트 코드를 바운디드 컨텍스트 별로 분리하는 게 좋아 보입니다.

HackSung reviewed (changes requested):
3단계 미션 잘 수행하시고 계신 것 같습니다. 👍 
정환님 질문에 대한 저의 생각과 몇가지 피드백을 남겨드렸으니 확인 부탁드립니다. 🙏

writer0713 committed:
refactor: 바운디드 컨텍스트 기준으로 패키지 수정
- 'eat_in_order', 'delivery_order', 'take_out_order' 에 각각 알맞는 OrderStatus 추가
- 'DIP' 를 사용하도록 구성 (application / domain / infra)
- 테스트코드에서 import 문 알맞게 수정

writer0713 commented:
@HackSung 
application / domain / infra 패키지에 어떤 클래스 (인터페이스) 를 위치시키는지에 대한 기준을 잘 모르고 있었던것 같습니다.
말씀주신 피드백을 보고 최범균님이 쓰신 `도메인 주도 개발 시작하기` 책을 보니 패키지를 구성하는 기준을 잘 설명하고 있더라구요.

책 보고 이해한것을 토대로 피드백 주신 부분 수정하여 push 했습니다.
감사합니다 :) 

(order 패키지 및 테스트코드는 추후 `도메인 주도 설계 기본 요소 - 리팩토링` 에서 진행하는 것 같아 따로 삭제 / 수정하지 않았습니다)

### 수정한 내용들
- 'eat_in_order', 'delivery_order', 'take_out_order' 에 각각 알맞는 상태를 가진 `OrderStatus` 추가
- 'DIP' 를 사용하도록 구성 (application / domain / infra)
  - interface 는 application 과 domain 패키지에 위치, 실제 구현체는 infra 에 위치
  
### package tree

```js
├── Application.java
├── common
│   ├── application
│   │   └── PurgomalumClient.java
│   └── infra
│       └── DefaultPurgomalumClient.java
├── delivery_order
│   ├── application
│   ├── domain
│   │   ├── DeliveryOrder.java
│   │   ├── DeliveryOrderRepository.java
│   │   └── OrderStatus.java
│   ├── infra
│   │   └── JpaDeliveryOrderRepository.java
│   └── ui
├── eat_in_order
│   ├── application
│   ├── domain
│   │   ├── EatInOrder.java
│   │   ├── EatInOrderRepository.java
│   │   └── OrderStatus.java
│   ├── infra
│   │   └── JpaEatInOrderRepository.java
│   └── ui
├── menu
│   ├── application
│   │   └── MenuService.java
│   ├── domain
│   │   ├── JpaMenuRepository.java
│   │   ├── Menu.java
│   │   ├── MenuProduct.java
│   │   └── MenuRepository.java
│   ├── infra
│   └── ui
│       └── MenuRestController.java
├── menu_group
│   ├── application
│   │   └── MenuGroupService.java
│   ├── domain
│   │   ├── JpaMenuGroupRepository.java
│   │   ├── MenuGroup.java
│   │   └── MenuGroupRepository.java
│   ├── infra
│   └── ui
│       └── MenuGroupRestController.java
├── order
│   ├── application
│   │   └── OrderService.java
│   ├── domain
│   │   ├── JpaOrderRepository.java
│   │   ├── Order.java
│   │   ├── OrderLineItem.java
│   │   ├── OrderRepository.java
│   │   ├── OrderStatus.java
│   │   └── OrderType.java
│   └── ui
│       └── OrderRestController.java
├── order_table
│   ├── application
│   │   └── OrderTableService.java
│   ├── domain
│   │   ├── JpaOrderTableRepository.java
│   │   ├── OrderTable.java
│   │   └── OrderTableRepository.java
│   ├── infra
│   └── ui
│       └── OrderTableRestController.java
├── product
│   ├── application
│   │   └── ProductService.java
│   ├── domain
│   │   ├── JpaProductRepository.java
│   │   ├── Product.java
│   │   └── ProductRepository.java
│   ├── infra
│   └── ui
│       └── ProductRestController.java
└── take_out_order
    ├── application
    │   └── KitchenridersClient.java
    ├── domain
    │   ├── OrderStatus.java
    │   ├── TakeOutOrder.java
    │   └── TakeOutOrderRepository.java
    ├── infra
    │   ├── DefaultKitchenridersClient.java
    │   └── JpaTakeOutOrderRepository.java
    └── ui

```

HackSung commented on src/main/java/kitchenpos/common/application/PurgomalumClient.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.infra;
+package kitchenpos.common.application;
```
application계층의 역할은 사용자의 요구사항을 처리하고 로직 수행을 domain 모델에 위임한다고 했을때
PurgomalumClient 인터페이스는 비속어 검사를 하는 도메인 로직에 해당하므로 domain계층에 위치하는게
더 적절해 보이는데 정환님의 생각은 어떠세요? 🤔

writer0713 replied:
사실 이전 pr 올릴때만해도 `application service` 와 `domain service` 의 차이점에 대해 잘 이해가 안갔었습니다.
이후에 최범균님 책 `도메인 주도 개발 시작하기` 에서 애플리케이션 서비스와 도메인 서비스의 차이에 대해 설명한걸 보고 아래와 같이 이해하게 됐습니다.

> 도메인의 상태를 변경하거나 도메인의 상태를 사용하는 서비스를 도메인 서비스로 볼수있다.

`PurgomalumClient` 의 경우, 도메인의 이름에 비속어가 있는지 검사하는 역할을 하니 도메인 서비스로 보는게 맞을것 같습니다 :)

HackSung commented on src/main/java/kitchenpos/order_table/domain/JpaOrderTableRepository.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.domain;
+package kitchenpos.order_table.domain;
```
도메인 계층에 아직 남아 있는 `Jpa...Repository`를 찾아서 모두 infra계층으로 옮겨 보는건 어떠세요?

writer0713 replied:
에고.. 꼼꼼하게 봤어야하는데 덤벙거렸네요 ^^;
모두 `infra` 계층으로 옮기겠습니다.

HackSung commented on src/main/java/kitchenpos/order/domain/OrderLineItem.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.domain;
+package kitchenpos.order.domain;
```
다음에 진행하실 `도메인 주도 설계 기본 요소` 미션들은 각 바운디드 컨텍스트별로 하나씩 리팩토링하는 미션이라
이번 미션에서 BOUNDED CONTEXT별 전체 패키지 구조를 잡고 가는게 좋을 것 같아요.
따라서 주문유형별로 독립적인 Bounded context를 구성하시고 order패키지 삭제까지 시도 해보는 건 어떠세요?

HackSung reviewed (changes requested):
미션 진행에 수고 많으셨습니다. 😄 
좀 더 시도해 보았으면 하는 피드백을 남겨 드렸으니 확인 부탁드려요 🙏

writer0713 committed:
refactor: Jpa..Repository 모두 infra 패키지로 이동
- 특정기술을 구현한 Jpa...Repository 는 infra 로 이동

writer0713 committed:
refactor: PurgomalumClient 도메인 패키지로 이동
- PurgomalumClient 는 도메인 서비스이기에 도메인 패키지로 이동한다

writer0713 committed:
refactor: 타입별 주문 (xxxOrder) 과 OrderLineItem 관계 재정의
- fk name 재정의
- joinColumn name 중복으로 인해 재정의
- 포장주문과 배달주문 repository 에서 existsByOrderTableAndStatusNot 메서드 제거

writer0713 committed:
feat: DeliveryOrderService 추가

writer0713 committed:
feat: EatInOrderService 추가

writer0713 committed:
feat: TakeOutOrderService 추가

writer0713 committed:
test: 테스트도 bounded context 별로 패키지 구성

writer0713 committed:
test: 주문배달 테스트 추가

writer0713 committed:
refactor: 배달주문, 매장주문 로직 수정
- 매장주문이 아닌 주문은 수량이 0보다 작을수 없다

writer0713 committed:
test: 포장주문 관련 테스트 추가

writer0713 committed:
feat: 실수로 지운 DefaultKitchenridersClient 추가

writer0713 committed:
test: 매장주문 테스트 추가

writer0713 committed:
refactor: order 패키지 정리
- order.domain 에서 공통으로 사용되는 OrderLineItem 과 OrderType 제외하고 모두 삭제
- 각 주문별 ui 패키지에 controller 추가

writer0713 committed:
refactor: order 패키지 내부로 매장주문, 배달주문, 포장주문 패키지를 이동

writer0713 commented:
@HackSung 
피드백 주신 부분들 수정하고 리팩토링 진행했습니다.
전체 패키지 구조 개선 및 테스트도 수정했습니다.

order 패키지 내부로 각 주문 패키지 (delivery_order, eat_in_order, take_out_order) 를 넣었습니다.
수정을 하다보니 order 패키지 내부에 `OrderLineItem` 과 `OrderType` 을 공통으로 사용하게되어 결과적으로 order 패키지를 상위로 하고 각 패키지가 내부로 들어오는 구조가 됐습니다.

```js
├── Application.java
├── common
│   ├── application
│   ├── domain
│   │   └── PurgomalumClient.java
│   └── infra
│       └── DefaultPurgomalumClient.java
├── menu
│   ├── application
│   │   └── MenuService.java
│   ├── domain
│   │   ├── Menu.java
│   │   ├── MenuProduct.java
│   │   └── MenuRepository.java
│   ├── infra
│   │   └── JpaMenuRepository.java
│   └── ui
│       └── MenuRestController.java
├── menu_group
│   ├── application
│   │   └── MenuGroupService.java
│   ├── domain
│   │   ├── MenuGroup.java
│   │   └── MenuGroupRepository.java
│   ├── infra
│   │   └── JpaMenuGroupRepository.java
│   └── ui
│       └── MenuGroupRestController.java
├── order
│   ├── delivery_order
│   │   ├── application
│   │   │   ├── DeliveryOrderService.java
│   │   │   └── KitchenridersClient.java
│   │   ├── domain
│   │   │   ├── DeliveryOrder.java
│   │   │   ├── DeliveryOrderRepository.java
│   │   │   └── OrderStatus.java
│   │   ├── infra
│   │   │   ├── DefaultKitchenridersClient.java
│   │   │   └── JpaDeliveryOrderRepository.java
│   │   └── ui
│   │       └── DeliveryOrderRestController.java
│   ├── domain
│   │   ├── OrderLineItem.java
│   │   └── OrderType.java
│   ├── eat_in_order
│   │   ├── application
│   │   │   └── EatInOrderService.java
│   │   ├── domain
│   │   │   ├── EatInOrder.java
│   │   │   ├── EatInOrderRepository.java
│   │   │   └── OrderStatus.java
│   │   ├── infra
│   │   │   └── JpaEatInOrderRepository.java
│   │   └── ui
│   │       └── EatInOrderRestController.java
│   └── take_out_order
│       ├── application
│       │   └── TakeOutOrderService.java
│       ├── domain
│       │   ├── OrderStatus.java
│       │   ├── TakeOutOrder.java
│       │   └── TakeOutOrderRepository.java
│       ├── infra
│       │   └── JpaTakeOutOrderRepository.java
│       └── ui
│           └── TakeOutOrderRestController.java
├── order_table
│   ├── application
│   │   └── OrderTableService.java
│   ├── domain
│   │   ├── OrderTable.java
│   │   └── OrderTableRepository.java
│   ├── infra
│   │   └── JpaOrderTableRepository.java
│   └── ui
│       └── OrderTableRestController.java
└── product
    ├── application
    │   └── ProductService.java
    ├── domain
    │   ├── Product.java
    │   └── ProductRepository.java
    ├── infra
    │   └── JpaProductRepository.java
    └── ui
        └── ProductRestController.java

```

HackSung commented on src/main/java/kitchenpos/order/delivery_order/application/KitchenridersClient.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.infra;
+package kitchenpos.order.delivery_order.application;
```
Application계층의 DeliveryOrderService가 KitcheridersClient를 의존하고 있는데
KitcheridersClient도 PurgomalumClient 처럼 domain계층으로 옮겨 보는 건 어떨까요?

writer0713 replied:
@HackSung 
에고.. 왜 항상 이렇게 피드백을 들어야 그떄서야 생각이 날까요.. ㅎㅎ
감사합니다 :)

HackSung commented:
작성하신 모델링에 의거하여 주문 유형별 바운디드 컨텍스트를 패키지로 잘 나누신 것 같습니다. 👍 

참고로 MenuGroup도메인은 Menu도메인에서만 참조하고 있고 
OrderTable도메인은 EatinOrder도메인에서만 사용하고 있으니
아래와 같은 형태로 해당 도메인을 같은 상위 패키지 내에 묶어 볼수 도 있습니다.
(반영은 안하셔도 됩니다. 다음 리팩터링 미션에서 한번 고민해 보시면 좋을 것 같아요 😄 ) 

```
├── menus
│   ├── domain
│   │   ├── menu
│   │   └── menugroup
...
...
├── order
│   ├── eat_in_order
│   │   ├── domain
│   │   │   ├── order
│   │   │   └── ordertable
```

HackSung reviewed (changes requested):
피드백 아주 잘 반영해 주신 것 같습니다. 👍 
사소한 피드백을 추가로 남겼으니 확인만 하시고 다시 PR주시면 바로 승인 및 머지하겠습니다.  😄

writer0713 committed:
refactor: KitchenridersClient 를 도메인 패키지로 이동
- KitchenridersClient 를 사용하는것도 도메인 속성을 사용하는 로직이라서 도메인 패키지로 이동

writer0713 committed:
refactor: menu_group 을 menu 패키지 내부로 이동
- menu_group 은 menu 와 항상 함께 사용된다.

writer0713 committed:
refactor: order_table 을 eat_in_order 패키지 하위로 이동
- order_table 은 eat_in_order 와 항상 함께 사용됨

writer0713 commented:
@HackSung 
꼼꼼한 피드백 너무 감사드립니다 :) 
말씀주신 부분들 반영했습니다.

- menu_group 을 menu 하위로 이동
- table_order 를 eat_in_order 하위로 이동
- KitchenridersClient 를 domain 패키지로 이동

꼼꼼하게 리뷰해주시고 여러가지 관점에서 생각해볼수 있게 도와주셔서 감사합니다 :)

HackSung reviewed (approved):
마지막 3단계 미션을 끝까지 마무리 하시느라 고생 많으셨습니다. 👏 
지금까지 열심히 해오셨던 것처럼 남은 미션도 완주하시길 응원합니다. 🙏
